Sources/CryptoBoringWrapper/EC/EllipticCurvePoint.swift

```diff
@@ -12,41 +12,262 @@
+            try group.withUnsafeGroupPointer { groupPtr in
+                guard CCryptoBoringSSL_EC_POINT_invert(groupPtr, selfPtr, nil) != 0 else {
                     throw CryptoBoringWrapperError.internalBoringSSLError()
                 }
             }
         }
     }
 
-    package init(copying pointer: OpaquePointer, on group: BoringSSLEllipticCurveGroup) throws {
+    @usableFromInline
+    package convenience init(inverting point: EllipticCurvePoint, on group: BoringSSLEllipticCurveGroup) throws {
+        try self.init(copying: point, on: group)
+        try self.invert(on: group)
+    }
+
+    @usableFromInline
+    package func inverting(on group: BoringSSLEllipticCurveGroup) throws -> EllipticCurvePoint {
+        try EllipticCurvePoint(inverting: self, on: group)
+    }
+
+    @usableFromInline
+    package static func inverting(_ point: EllipticCurvePoint, on group: BoringSSLEllipticCurveGroup) throws -> EllipticCurvePoint {
+        try EllipticCurvePoint(inverting: point, on: group)
+    }
+
+    @usableFromInline
+    package func subtract(_ rhs: EllipticCurvePoint, on group: BoringSSLEllipticCurveGroup) throws {
+        try self.add(rhs.inverting(on: group), on: group)
+    }
+
+    @usableFromInline
+    package convenience init(subtracting rhs: EllipticCurvePoint, from lhs: EllipticCurvePoint, on group: BoringSSLEllipticCurveGroup) throws {
+        try self.init(copying: lhs, on: group)
+        try self.subtract(rhs, on: group)
+    }
+
+    @usableFromInline
+    package func subtracting(_ rhs: EllipticCurvePoint, on group: BoringSSLEllipticCurveGroup) throws -> EllipticCurvePoint {
+        try EllipticCurvePoint(subtracting: rhs, from: self, on: group)
+    }
+
+    @usableFromInline
+    package static func subtracting(_ rhs: EllipticCurvePoint, from lhs: EllipticCurvePoint, on group: BoringSSLEllipticCurveGroup) throws -> EllipticCurvePoint {
+        try EllipticCurvePoint(subtracting: rhs, from: lhs, on: group)
+    }
+
+    @usableFromInline
+    package init<MessageBytes: ContiguousBytes, DSTBytes: ContiguousBytes>(hashing msg: MessageBytes, to group: BoringSSLEllipticCurveGroup, domainSeparationTag: DSTBytes) throws {
+        let hashToCurveFunction = switch group.curveName {
+        case .p256: CCryptoBoringSSLShims_EC_hash_to_curve_p256_xmd_sha256_sswu
+        case .p384: CCryptoBoringSSLShims_EC_hash_to_curve_p384_xmd_sha384_sswu
+        case .p521: throw CryptoBoringWrapperError.invalidParameter // BoringSSL doesn't have a hash_to_curve API for P521.
+        case .none: throw CryptoBoringWrapperError.internalBoringSSLError()
+        }
+
+       self._basePoint = try msg.withUnsafeBytes { msgPtr in
+            try group.withUnsafeGroupPointer { groupPtr in
+                try domainSeparationTag.withUnsafeBytes { dstPtr in
+                    guard
+                        let pointPtr = CCryptoBoringSSL_EC_POINT_new(groupPtr),
+                        hashToCurveFunction(
```

Comment:
If this fails, we leak the point.

Reply:
Resolved in the same manner as https://github.com/apple/swift-crypto/pull/292/files/fda1a2c6df294e87b24b77ad28a7cd5950d74d25#r1850181741.